include/ginkgo/core/distributed/vector.hpp

```diff
@@ -510,34 +510,28 @@ class Vector
 }  // namespace experimental
 
 
-template <typename ValueType>
-struct polymorphic_object_traits<experimental::distributed::Vector<ValueType>> {
-    using Vector = experimental::distributed::Vector<ValueType>;
+namespace detail {
 
-    static std::unique_ptr<PolymorphicObject> create_default_impl(
-        const Vector* self, std::shared_ptr<const Executor> exec)
-    {
-        return std::unique_ptr<Vector>{
-            new Vector(exec, self->get_communicator())};
-    }
 
-    static std::unique_ptr<Vector> create_conversion_target_impl(
-        const experimental::distributed::Vector<next_precision<ValueType>>*
-            self,
-        std::shared_ptr<const Executor> exec)
-    {
-        return std::unique_ptr<Vector>{
-            new Vector(exec, self->get_communicator())};
-    }
+template <typename TargetType>
+struct conversion_target_helper;
+
+
+template <typename ValueType>
+struct conversion_target_helper<experimental::distributed::Vector<ValueType>> {
+    using target_type = experimental::distributed::Vector<ValueType>;
+    using source_type =
+        experimental::distributed::Vector<gko::previous_precision<ValueType>>;
 
-    static PolymorphicObject* clear_impl(Vector* self)
+    static std::unique_ptr<target_type> create(
+        const source_type* source, std::shared_ptr<const Executor> exec)
     {
-        *self = Vector{self->get_executor(), self->get_communicator()};
-        return self;
+        return target_type::create(exec, source->get_communicator());
     }
 };
```

Comment:
Why is this restricted to only the `source_type`? I don't see any reason why `source_type` couldn't just be a template for `create` (a name I would like to be changed, see previous comment) as it is usually.
If you don't need it, `previous_precision` can also be removed (as this is the only use case for it at the moment).
If you want / need to keep it here, please remove the `gko::` namespace prefix from it here.

Reply:
Because the vector only supports converting to the next precision. So every other type wouldn't make sense to use here, and by being explicit with the type, a mismatch can be caught at compile time,

Reply:
Don't you have the same issue with your default implementation of `convresion_target_helper`? It is a bit strange to me that the default implementation can handle pretty much all cases, while the `distributed::Vector` implementation only handles specific `ValueType` combinations.
Additionally, since we only have 2 different `ValueType`s here, it makes no difference (and would actually remove the need for `previous_precision`).

Reply:
In the general implementation, we have no idea, which SourceTypes can be converted into the TargetType, but for the special case, we know exactly for which types this is possible. I think we should use that additional information to make it more type safe.
I get that the `previous_precision` is controversial, but for me having the source type as the next precision, while our conversions are usually from lower to higher precision, seems confusing. Also, each vector type is only convertible to `next_precision<value_type>`, so getting the type that a vector can be a target of a conversion would mean to me that we need the inverse of the `next_precision` operation. I tried to add that with the `previous_precision`.

Reply:
I don't really have an issue with `previous_precision` as this is indeed useful to have.
It is just strange to me that we have one implementation where you can use it with pretty much any `SourceType`, and one that heavily restricts it. In theory, you could check `SourceType::value_type` and see if it fulfills your requirement, but I understand that this is also not ideal.
I am fine if you want to keep it that way, but please add a short comment why the restriction exists in that case.

Reply:
It's actually quite simple to restrict the SourceType by using SFINAE with `std::is_base_of<ConvertibleTo<TargetType>, SourceType>`. I will add that in the general case, and leave the vector specialization as it is. Thanks for bringing the whole issue up!